performance_manager/lib/rt_vehicle_positions.py

```diff
@@ -0,0 +1,277 @@
+    df["timestamp_end"] = numpy.where(
+        (df["hash"] - df["hash"].shift(-1) == 0),
+        df["timestamp_start"].shift(-1),
+        df["timestamp_start"],
+    ).astype("int64")
+
+    """
+    for matching consectuive hash events, drop 2nd event because timestamp
+    has been captured in timestamp_end value of previous event
+    """
+    drop_last_mask = df["hash"] - df["hash"].shift(1) != 0
+    df = df.loc[drop_last_mask, :].reset_index(drop=True)
+
+    df.insert(0, "id", numpy.nan)
+
+    # fill all dataframe na values with Python None (for DB writing)
+    df = df.fillna(numpy.nan).replace([numpy.nan], [None])
+
+    return df
+
+
+def get_utc_timestamp(path: str) -> float:
+    """
+    process datetime from s3 path return UTC timestamp
+    """
+    year = int(re.findall(r"year=(\d{4})", path)[0])
+    month = int(re.findall(r"month=(\d{1,2})", path)[0])
+    day = int(re.findall(r"day=(\d{1,2})", path)[0])
+    hour = int(re.findall(r"hour=(\d{1,2})", path)[0])
+    dt = datetime.datetime(
+        year=year, month=month, day=day, hour=hour, tzinfo=datetime.timezone.utc
+    )
+    return datetime.datetime.timestamp(dt)
+
+
+def merge_vehicle_position_events(
+    folder: str,
+    new_events: pandas.DataFrame,
+    session: sqlalchemy.orm.session.sessionmaker,
+) -> None:
+    """
+    merge new events from parquet file with exiting events found in database
+    new events will be merged with existing events in a 5 minutes window
+    surrounding the year/month/day/hour value found in path of parquet files
+    """
+
+    start_window_ts = get_utc_timestamp(folder)
+    timestamp_to_pull_min = start_window_ts - 300
+    timestamp_to_pull_max = start_window_ts + 300 + 3600
+    get_db_events = select(
+        (
+            VehiclePositionEvents.id,
+            VehiclePositionEvents.hash,
+            VehiclePositionEvents.timestamp_start,
+            VehiclePositionEvents.timestamp_end,
+            VehiclePositionEvents.vehicle_id,
+        )
+    ).where(
+        (VehiclePositionEvents.timestamp_start > timestamp_to_pull_min)
+        & (VehiclePositionEvents.timestamp_end < timestamp_to_pull_max)
+    )
+    with session.begin() as s:  # type: ignore
+        if s.execute(get_db_events).first() is None:
+            logging.info(f"No merge events found in database.")
+            merge_events = new_events.sort_values(
+                by=["vehicle_id", "timestamp_start"]
+            )
+        else:
+            # Join records from db with records from parquet
+            # db records should contain 'index' column that does not exist in parquet dataset
+            merge_events = pandas.concat(
+                [
+                    pandas.DataFrame(
+                        [r._asdict() for r in s.execute(get_db_events)]
+                    ),
+                    new_events,
+                ]
+            ).sort_values(by=["vehicle_id", "timestamp_start"])
```

Comment:
Pull existing records for DB that are inside of the merge window and create a `merge_events` sorted dataframe with new and existing records.